src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -163,16 +208,24 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_copy_ep_get_short,
                  uct_ep_h tl_ep, void *buffer, unsigned length,
                  uint64_t remote_addr, uct_rkey_t rkey)
 {
-    uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_cuda_copy_iface_t);
-    uct_cuda_copy_stream_t idx   = UCT_CUDA_COPY_STREAM_D2H;
+    uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_ep->iface,
+                                                  uct_cuda_copy_iface_t);
+    ucs_memory_type_t src_type;
+    ucs_memory_type_t dst_type;
+    cudaStream_t *stream;
     ucs_status_t status;
 
-    UCT_CUDA_COPY_CHECK_AND_CREATE_STREAM(iface, idx);
+    uct_cuda_copy_get_mem_types((const void*)remote_addr, (const void*)buffer,
```

Comment:
hmm, this means two memory pointer detection overhead in a small message zcopy/bcopy protocol?  this could be significant ~1us overhead for ~7us 8 byte latency

Reply:
this is blocking API, why do we care about memory type-specific stream? can we use one dedicated stream here?

Reply:
Good point. I'll add a dedicated stream.

Reply:
And for zcopy cases also, can we option to avoid memory type detection as in most of the cases there might single type CUDA memory (cuda or managed) type might be in use. we might be good with just a stream per direction in 99 % of cases.
even if this not default, it might be useful to have an option to avoid this memory detection. WDYT?

Reply:
@bureddy 
It's 99% cases. For alltoall over managed memory example I gave will have m2h/h2m and m2d/d2m transfers for inter-node and intra-node concurrently. We don't want serialization here. There are other examples without GDR where would use host and device bounce buffers concurrently for inter-node and intra-node transfers respectively.

For zero copy, the edge case is when transfer time d2d or d2h which takes 2us and query time is 0.2 * 2 us (because average attribute query cost is 0.2 us and not 0.5us). When we move to unified cache in the transport layer, this cost will come down further. Without detection, I don't see a way of avoiding serialization. Do you have ideas how we may get around it?

Reply:
@bureddy Let me know if changes in https://github.com/openucx/ucx/pull/6115/commits/e7d1a4f9bf843ea17af752d7d108000fa28460f3 looks good for short operations.

Reply:
@bureddy 
> even if this not default, it might be useful to have an option to avoid this memory detection. WDYT?

Thinking about this a bit more, I had previous prototype where I created a separate sub-transport called managed_copy which would avoid detection approach used here. Should we go through that? It's a lot more code because we would need to add sub-transports for many paths ((m2h, h2m), (m2d, d2m), (d2d)). Also this is not scalable because we want start supporting new memory types like [cumemap](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VA.html#group__CUDA__VA) and [stream-allocated-memory](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html#group__CUDART__MEMORY__POOLS). It's better to have detection and handle things without replicating code. 

Also, discussing with @SeyedMir , being able to query memory type in transport layer through unified cache is becoming more and more relevant for different components. If that is the eventual solution, should we not just defer the memory detection optimization to a separate PR and just address the serialization problem here?

cc @yosefe

Reply:
I dont think separate tl is efficient. One another way to have a common stream pool and get a stream for each async copy and put it back.  This way, you dont care about type of the memory.  No.of streams needed will depend on the no.of outstanding copy ops. I'm not sure what is the overhead of creating a pool of streams